test/ns_test.go

```diff
@@ -0,0 +1,132 @@
+package telepilot_test
+
+import (
+	"io"
+	"os"
+	"strings"
+	"testing"
+
+	pb "go.creack.net/telepilot/api/v1"
+)
+
+func TestPIDNamespace(t *testing.T) {
+	t.Parallel()
+
+	ts, ctx := newTestServer(t)
+
+	t.Run("ps", func(t *testing.T) {
+		t.Parallel()
+
+		jobID, err := ts.alice.StartJob(ctx, "ps", []string{"-e"})
+		noError(t, err, "Start job.")
+		t.Cleanup(func() { noError(t, ts.alice.StopJob(ctx, jobID), "Cleanup stop job.") })
+
+		w := &strings.Builder{}
+		noError(t, ts.alice.StreamLogs(ctx, jobID, w), "Stream logs.")
+		// We expect only 2 lines: the headers and the lone process in the namesapce.
+		if l := len(strings.Split(strings.TrimSpace(w.String()), "\n")); l != 2 {
+			t.Fatalf("Unexpected line count for `ps -e`: %d.\n%s\n", l, w)
+		}
+	})
+	t.Run("getpid", func(t *testing.T) {
+		t.Parallel()
+
+		jobID, err := ts.alice.StartJob(ctx, "sh", []string{"-c", "echo $$"})
+		noError(t, err, "Start job.")
+		t.Cleanup(func() { noError(t, ts.alice.StopJob(ctx, jobID), "Cleanup stop job.") })
+
+		w := &strings.Builder{}
+		noError(t, ts.alice.StreamLogs(ctx, jobID, w), "Stream logs.")
+		// We expect to be pid 1.
+		assert(t, "1", strings.TrimSpace(w.String()), "invalid pid")
+	})
+}
+
+func TestNetworkNamespace(t *testing.T) {
+	t.Parallel()
+
+	ts, ctx := newTestServer(t)
+
+	t.Run("iface", func(t *testing.T) {
+		t.Parallel()
+
+		jobID, err := ts.alice.StartJob(ctx, "ip", []string{"address", "show"})
+		noError(t, err, "Start job.")
+		t.Cleanup(func() { noError(t, ts.alice.StopJob(ctx, jobID), "Cleanup stop job.") })
+
+		w := &strings.Builder{}
+		noError(t, ts.alice.StreamLogs(ctx, jobID, w), "Stream logs.")
+		// We expect only 1 entry of 2 lines: the loopback.
+		if l := len(strings.Split(strings.TrimSpace(w.String()), "\n")); l != 2 {
+			t.Fatalf("Unexpected iface count: %d\n", l)
+		}
+		if !strings.Contains(w.String(), "1: lo:") {
+			t.Fatal("Invalid network interface output. No loopback.")
+		}
+	})
+	t.Run("ping", func(t *testing.T) {
+		t.Parallel()
+
+		// Attempt to ping Cloudflare.
+		// NOTE: Depending on the system, ping may exit different codes. Wrap it in a shell to control
+		// the exit code.
+		jobID, err := ts.alice.StartJob(ctx, "sh", []string{"-c", "ping 1.1.1.1 -c 1 || exit 1"})
+		noError(t, err, "Start job.")
+		t.Cleanup(func() { noError(t, ts.alice.StopJob(ctx, jobID), "Cleanup stop job.") })
+
+		w := &strings.Builder{}
+		noError(t, ts.alice.StreamLogs(ctx, jobID, w), "Stream logs.")
+
+		st, err := ts.alice.GetJobStatus(ctx, jobID)
+		noError(t, err, "Get job status.")
+		assert(t, pb.JobStatus_JOB_STATUS_EXITED.String()+" (1)", st, "invalid status")
+	})
+}
+
+func TestMountNamespace(t *testing.T) {
+	t.Parallel()
+
+	ts, ctx := newTestServer(t)
+
+	pipeName := mkTempPipe(t, "pipe")
+
+	mountPoint := t.TempDir()
```

Comment:
It's worth noting that, with "true" isolation, this wouldn't be a path on the job filesystem.

Reply:
Added comment about it.